### PR TITLE
Handle set-cookie from a redirect response fixes #13

### LINKED ISSUE #13: Handle set-cookie for redirects
Seems like we need to handle cookies that are set when the response is `301, 302, 303, 307, 308`.

e.g. 302 responses can contain `response.headers.set-cookie` that we need to store.

To implement it `fetch` needs to handle redirects manually:
```ts
...
const response = await fetch(input, {
  ...interceptedInit,
  redirect: "manual",
});
...
```

... and check if it is a redirect status in `response.status`, then if it's a redirect; recursively call it self with something like.
```ts
...
    const redirectStatus = new Set([301, 302, 303, 307, 308]);
    if (!redirectStatus.has(response.status)) return response

    const redirectUrl = response.headers.get("location");
    if (!redirectUrl) return response;

    return await wrappedFetch(redirectUrl, init);
...
```

We also need to add a `maxRedirect` option to check against a counter to avoid redirect loops.

I don't have the time to finish this now, but started a draft that i think will work. Maybe someone will pick it up :-)

### PR DESCRIPTION
**Pull request by maccyber**
TODO:

- [x] Add maxRedirect, redirectCount
- [x] Test better
- [x] Handle if `redirect: "manual"` is sent through fetch options
- [x] Handle when redirected if `redirect: "error"` is sent through fetch options
- [x] Set `response.redirected` when redirected
- [x]  deepStrictEqual assert on user's init so we make sure to not accidentally mutate their init
- [x] Do not forward sensitive headers to third-party domains

### CODE REVIEW

File: fetch_wrapper.ts [hunk 1]
```diff
@@ -37,13 +46,53 @@ export function wrapFetch(options?: WrapFetchOptions): typeof fetch {
     }
     interceptedInit.headers.set("cookie", cookieString);
 
-    const response = await fetch(input, interceptedInit);
+    const response = await fetch(input, {
```
Comment (jd1378) [inline fetch_wrapper.ts]: can we not use spread operator for this and simply set `redirect` on `interceptedInit` directly ?
the reason is spread has a cost, and I prefer to not use it when It's not needed
even if we want to clone user's `init`, in this case which we should have, let's do it once and when we assign the `interceptedInit`
also probably change `interceptedInit` type to `WrappedFetchRequestInit` ?

File: fetch_wrapper.ts [hunk 2]
```diff
@@ -37,13 +46,53 @@ export function wrapFetch(options?: WrapFetchOptions): typeof fetch {
...
+    const maxRedirect = init?.maxRedirect ?? 20;
+
+    if (redirectCount >= maxRedirect) {
+      await response.body?.cancel();
+      throw new TypeError(
+        `Reached maximum redirect of ${maxRedirect} for URL: ${response.url}`,
+      );
+    }
+
+    init = {
```
Comment (jd1378) [inline fetch_wrapper.ts]: same here about the spread
also shouldn't it be changing `interceptedInit` ?

**Comment by jd1378** [comment]
also maybe do a deepStrictEqual assert on user's init so we make sure to not accidentally mutate their init from now on ?

File: fetch_wrapper.ts (hunk 1 above)
Comment (maccyber) [reply fetch_wrapper.ts]: Agreed. Tried to refactor it, but it still doesn't feel right. Must be better ways to solve it.
I hope someone else can push this over the line.

Comment (jd1378) [reply fetch_wrapper.ts]: thanks for the effort
I agree, it does not feel right.
I'll work on it when I have some time

**Comment by jd1378** [comment]
I have committed some changes
the reason for deciding to not handle Request was that it was not following whatwg fetch specs
so I thought its better to only replace headers and redirect using our init instead and let fetch handle the Request object

also I think the naming was good the way it was, so I renamed them back, sorry

the only thing remaining from tasks is `Handle deletion of cookie headers` which I don't quite understand
can you explain it a bit more ?

**Comment by jd1378** [comment]
I'm ready to merge, the only remaining task which I did not understand was "Handle deletion of cookie headers"
CookieJar already deletes a cookie if an expired cookie is set, so I think that shouldn't be a problem
please explain so we can merge this

**Comment by maccyber** [comment]
Awesome. I updated the task to "Do not forward sensitive headers (i.e. `headers.authorization` and `headers.cookie`) to third-party domains" to clearify.

I think I can create a test for it later tonight

**Comment by jd1378** [comment]
since I still did not understand, I appreciate the incoming tests :D
Links/examples in this case for the expected behavior would be appreciated
my first guess is to add support for `SameSite` attribute of cookies ?
and then I guess maybe what you suggest is a privacy feature ? for more strictness ?
I would be happy to help on this as well

edit: my bad, `SameSite` attribute is unrelated, since its only happening in browser environment
edit 2: is this related to redirection feature ? if so, the fetch should handle it using `credentials` option in `init` I believe ?

**Comment by maccyber** [comment]
Implementation of this functionality in fetch-cookie: https://github.com/valeriangalliat/fetch-cookie/blob/master/src/index.ts#L161-L166

**Comment by jd1378** [comment]
from what I understood, we were already hitting the check [here](https://github.com/jd1378/deno-another-cookiejar/blob/9dd180a3b5c9a0a1e74f00c14a37c40dd18eccc2/cookie_jar.ts#L204) and resetting it always [here](https://github.com/jd1378/deno-another-cookiejar/blob/9dd180a3b5c9a0a1e74f00c14a37c40dd18eccc2/fetch_wrapper.ts#L24)

so I think adding a test should make it clear if It's as I'm thinking or we need to change anything

edit: sorry, I keep missing the point
I understood now whats happening
the "headers" not just "cookie" header
my bad again

**Comment by jd1378** [comment]
I would be happy if you took a look at the test and the code guarding what it should

the cookie header part is handled differently by resetting them for every request

if everything is fine, tell me so I can merge
if you want to add more tests, you are very welcome

**Comment by jd1378** [comment]
I've also noticed I should have used `hostname` instead of `host` of URL object, because RFC 6265 under [8.5. Weak Confidentiality](https://www.rfc-editor.org/rfc/rfc6265#section-8.5)
So I also made that change to the code and tests and made server two use "localhost" vs "127.0.0.1" to differentiate them instead of relying on ports

**Comment by maccyber** [comment]
Nice, looks good to me

**Comment by jd1378** [comment]
thanks a lot for this !

**Pull request merged**